File: Labirinto.java
import java.util.ArrayList;

public class Labirinto {

    int largura, altura;
    Coordenada entrada, saida;
    char[][] labirinto;
    boolean quebrado = false;

    public Labirinto(int dificuldade, String tipo) throws Exception {
        if(dificuldade < 1)
            dificuldade = 1;

        largura = dificuldade * 4 - 1;
        altura  = dificuldade * 2 + 1;

        labirinto = new char[altura][largura];

        for(int i=0; i < altura; i++)
            for(int j=0; j < largura; j++)
                labirinto[i][j] = '#';

        switch (tipo) {
            case "fixo" -> {
                entrada = new Coordenada(largura / 2, 0);
                saida = new Coordenada(largura / 2, altura - 1);
            }
            case "randomico" -> {
                entrada = new Coordenada((int) (Math.random() * (largura / 2)) * 2 + 1, 0);
                saida = new Coordenada((int) (Math.random() * (largura / 2)) * 2 + 1, altura - 1);
            }
            case "quebrado" -> {
                entrada = new Coordenada((int) (Math.random() * (largura / 2)) * 2 + 1, 0);
                saida = new Coordenada((int) (Math.random() * (largura / 2)) * 2 + 1, altura - 1);
                quebrado = true;
            }
            default -> {
                entrada = new Coordenada(1, 0);
                saida = new Coordenada(largura - 2, altura - 1);
            }
        }

        Pilha<Coordenada> caminho = new Pilha<>(largura * altura);
        Coordenada atual = new Coordenada(entrada);

        caminho.guardeUmItem(atual);
        labirinto[entrada.getY()][entrada.getX()] = 'E'; //Abrir a entrada
        atual = new Coordenada(entrada.getX(),entrada.getY()+1);
        caminho.guardeUmItem(atual);
        labirinto[atual.getY()][atual.getX()] = ' ';

        do
        {
            ArrayList<Coordenada> fila = new ArrayList<>(3);

            if(cavavel(atual.getX() - 1, atual.getY()))
                fila.add(new Coordenada(atual.getX() - 1, atual.getY()));

            if(cavavel(atual.getX() + 1, atual.getY()))
                fila.add(new Coordenada(atual.getX() + 1, atual.getY()));

            if(cavavel(atual.getX(), atual.getY() - 1))
                fila.add(new Coordenada(atual.getX(), atual.getY() - 1));

            if(cavavel(atual.getX(), atual.getY() + 1))
                fila.add(new Coordenada(atual.getX(), atual.getY() + 1));

            for(;;)
            {
                if(fila.isEmpty())
                    break;

                int escolha = (int) (Math.random() * fila.size());

                Coordenada proximo = new Coordenada(atual.getX() + (fila.get(escolha).getX()-atual.getX()) * 2,
                                                    atual.getY() + (fila.get(escolha).getY()-atual.getY()) * 2);

                if(labirinto[proximo.getY()][proximo.getX()] == '#')
                    //Se a coordenada a 2 casa de distancia da direcao escolhida for parede ele cava para essa direcao
                {
                    cavar(fila.get(escolha));
                    cavar(proximo);

                    caminho.guardeUmItem(fila.get(escolha));
                    caminho.guardeUmItem(proximo);

                    atual = proximo;
                    break;
                }
                fila.remove(escolha);
            }

            if (fila.isEmpty())//Regressao
            {
                caminho.removaUmItem();
                caminho.removaUmItem();
                if(caminho.isVazia())
                    break;

                atual = caminho.recupereUmItem();
            }
        }while (true);
        labirinto[saida.getY()][saida.getX()] = 'S';
        if(quebrado) quebrar();
    }

    public boolean cavavel(int x, int y) {
        if(x <= 0 || x >= largura-1) return false;
        if(y <= 0 || y >= altura-1)  return false;
        return labirinto[y][x] == '#';
    }

    public void cavar(Coordenada cord) {
        labirinto[cord.getY()][cord.getX()] = ' ';

    }

    public void quebrar() { //Este metodo gera caminhos alternativos, já que só existe um caminho por padrao.
        for(int i = 0; i < altura; i++)
        {
            byte quebravel = 0;
            int x = (int) (Math.random() * (largura/ 2) - 1) * 2 + 2;
            int y = (int) (Math.random() * (altura / 2) - 1) * 2 + 2;

            if(labirinto[y][x-1] == '#') quebravel++;
            if(labirinto[y][x+1] == '#') quebravel++;
            if(labirinto[y-1][x] == '#') quebravel++;
            if(labirinto[y+1][x] == '#') quebravel++;

            if(quebravel == 2 && (labirinto[y][x-1] == '#' && labirinto[y][x+1] == '#' || labirinto[y-1][x] == '#' && labirinto[y+1][x] == '#'))
            {
                labirinto[y][x] = ' ';
                continue;
            }
            i--;
        }
    }

    public char[][] getLabirinto() {
        return labirinto;
    }

    @Override
    public String toString() {
        StringBuilder res = new StringBuilder();
        res.append(this.altura).append("\n").append(this.largura).append("\n");
        for(int i=0; i < altura; i++)
        {
            for(int j=0; j < largura; j++)
            {
                res.append(this.labirinto[i][j]);
            }
            res.append("\n");
        }
        return res.toString();
    }
}

File: Labirinto.cs
using System;
using System.Collections;

class Labirinto
{
    private int largura = 0;
    private int altura = 0;
    private readonly Coordenada entrada, saida;
    byte[,] matriz;
    private readonly bool quebrado = false;
    private readonly Random rand = new Random();

    public byte[,] Matriz => matriz;

    public int Largura => largura;
    public int Altura => altura;

    public Labirinto(int dificuldade, string tipo)
    {
        if (dificuldade < 1)
            dificuldade = 1;

        largura = dificuldade * 4 - 1;
        altura = dificuldade * 2 + 1;

        matriz = new byte[altura, largura];

        for (int i = 0; i < altura; i++)
            for (int j = 0; j < largura; j++)
                matriz[i, j] = 1;

        switch (tipo)
        {
            case "fixo":
                entrada = new Coordenada(largura / 2, 0);
                saida = new Coordenada(largura / 2, altura - 1);
                break;

            case "randomico":
                entrada = new Coordenada(rand.Next(largura / 2) * 2 + 1, 0);
                saida = new Coordenada(rand.Next(largura / 2) * 2 + 1, altura - 1);
                break;

            case "quebrado":
                entrada = new Coordenada(rand.Next(largura / 2) * 2 + 1, 0);
                saida = new Coordenada(rand.Next(largura / 2) * 2 + 1, altura - 1);
                quebrado = true;
                break;

            default:
                entrada = new Coordenada(1, 0);
                saida = new Coordenada(largura - 2, altura - 1);
                break;
        }

        Pilha<Coordenada> caminho = new Pilha<Coordenada>(largura * altura);
        Coordenada atual = new Coordenada(entrada);

        caminho.Empilhar(atual);
        matriz[entrada.Y, entrada.X] = 2; //Abrir a entrada
        atual = new Coordenada(entrada.X, entrada.Y + 1);
        caminho.Empilhar(atual);
        matriz[atual.Y, atual.X] = 0;

        do
        {
            ArrayList fila = new ArrayList(3);

            if (Cavavel(atual.X - 2, atual.Y))
                fila.Add(new Coordenada(atual.X - 2, atual.Y));

            if (Cavavel(atual.X + 2, atual.Y))
                fila.Add(new Coordenada(atual.X + 2, atual.Y));

            if (Cavavel(atual.X, atual.Y - 2))
                fila.Add(new Coordenada(atual.X, atual.Y - 2));

            if (Cavavel(atual.X, atual.Y + 2))
                fila.Add(new Coordenada(atual.X, atual.Y + 2));

            if (fila.Count <= 0)//Regressao
            {
                caminho.Desempilhar();
                caminho.Desempilhar();
                if (caminho.EstaVazia)
                    break;

                atual = caminho.OTopo();
                continue;
            }

            int escolha = rand.Next(fila.Count);

            Coordenada entre = new Coordenada((atual.X + ((Coordenada)fila[escolha]).X) / 2,
                                                            (atual.Y + ((Coordenada)fila[escolha]).Y) / 2);
            Cavar(entre);
            Cavar((Coordenada)fila[escolha]);

            caminho.Empilhar(entre);
            caminho.Empilhar((Coordenada)fila[escolha]);

            atual = (Coordenada)fila[escolha];

        } while (true);

        matriz[saida.Y, saida.X] = 3;

        if (quebrado)
            Quebrar();
    }

    public bool Cavavel(int x, int y)
    {
        try { if (x <= 0 || x >= largura - 1) return false; }
        catch (IndexOutOfRangeException) { return false; }
        try { if (y <= 0 || y >= altura - 1) return false; }
        catch (IndexOutOfRangeException) { return false; }
        return matriz[y,x] == 1;
    }

    public void Cavar(Coordenada cord)
    {
        matriz[cord.Y, cord.X] = 0;
    }

    public void Quebrar()
    {
        for (int i = 0; i < altura; i++)
        {
            byte quebravel = 0;
            int x = rand.Next((largura - 2) / 2) * 2 + 2;
            int y = rand.Next((altura - 2) / 2) * 2 + 2;

            if (matriz[y, x - 1] == 1) quebravel++;
            if (matriz[y, x + 1] == 1) quebravel++;
            if (matriz[y - 1, x] == 1) quebravel++;
            if (matriz[y + 1, x] == 1) quebravel++;

            if (quebravel == 2
                && (matriz[y, x - 1] == 1 && matriz[y, x + 1] == 1 || matriz[y - 1, x] == 1 && matriz[y + 1, x] == 1))
            {
                matriz[y, x] = 0;
                continue;
            }
            i--;
        }
    }

    public override string ToString()
    {
        string res = altura + "\n" + largura;
        for (int i = 0; i < altura; i++)
        {
            for (int j = 0; j < largura; j++)
            {
                res += this.matriz[i, j];
            }
            res += "\n";
        }
        return res;

    }
}
